### PR TITLE
MSC3015: Room state personal overrides

### PR DESCRIPTION
**Pull request by MurzNN**
[Rendered](https://github.com/MurzNN/matrix-doc/blob/personal-room-names/proposals/3015-room-personal-overrides.md)

Related issues:
- [Personal ('pet name') room name for chats](https://github.com/vector-im/element-meta/issues/725)
- [Ability to override the displayname for users & rooms that get served to you (SPEC-427)](https://github.com/matrix-org/matrix-doc/issues/690)

Signed-off-by: Alexey Korepov <MurzNN@gmail.com>

### CODE REVIEW

File: proposals/3015-room-personal-name.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# MSC3015: Room personal name
+
+Very often users want to personally rename rooms to see it in list like they wants, especially for DM rooms. But room name is shared thing, so if they rename it for yourself, all other members will see this rename too.
```
Comment (turt2live) [inline proposals/3015-room-personal-name.md:3]: please line wrap at about 120 characters for legibility.

Comment (MurzNN) [reply proposals/3015-room-personal-name.md]: Thanks for suggestion, done!

File: proposals/3015-room-personal-name.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
+# MSC3015: Room personal name
+
+Very often users want to personally rename rooms to see it in list like they wants, especially for DM rooms. But room
+name is shared thing, so if they rename it for yourself, all other members will see this rename too. 
+
+Examples of problem: 
+
+1. Very often other people want to rename DM rooms with me and can do this, because we both are admins. So he set the
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:8]:
```suggestion
1. Very often other people want to rename DM rooms with me and can do this, because we both are admins. So they set the
```

File: proposals/3015-room-personal-name.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+   alternative name of that DM room (eg `Korepov Alexey` instead of `Alexey Murz Korepov`) - this works well on his
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:9]:
```suggestion
   alternative name of that DM room (eg `Korepov Alexey` instead of `Alexey Murz Korepov`) - this works well on their
```

File: proposals/3015-room-personal-name.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+   side. But, as result on my side, I see that room in my rooms list with my name, instead of remote user name.
+
+2. The user have two DM rooms with different Matrix users, but both have name "Alice" without avatar. As result he see
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:12]:
```suggestion
2. The user has two DM rooms with different Matrix users, but both have name "Alice" without avatar. As result they see
```

File: proposals/3015-room-personal-name.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+   two identical rooms in list with same name and can understand which Alice is where only via reading recent messages.
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:13]:
```suggestion
   two identical rooms in the list with same name and it is unclear which room is which.
```

File: proposals/3015-room-personal-name.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+   And he can solve this problem via renaming rooms, that give the result from item 1, descibed above.
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:14]:
```suggestion
   If attempting to solve this problem via renaming rooms, the problem described in 1 occurs.
```

File: proposals/3015-room-personal-name.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+3. This problem often happens for rooms from bridged networks, when we talk with same person via different networks, and
+   want to mark each room personally. This can be solved via adding suffixes with remote network name to room name on
+   Bridge side, but people want to change other parts of room name too. For example, one person can have different names
+   in each network (eg "Alexey Korepov" on VK, "Korepov Alexey" on Skype, "Alex" on WhatsApp, "Murz" on Telegram), and I
+   want to see all rooms with him as similar names.
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:20]:
```suggestion
   want to see all the rooms with similar names.
```

File: proposals/3015-room-personal-name.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+Most of other modern messengers (Telegram, Skype, Viber, WhatsApp) already have this feature, but only for DM rooms via
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:22]:
```suggestion
Most other modern messengers (Telegram, Skype, Viber, WhatsApp) already have this feature, but only for DM rooms via
```

File: proposals/3015-room-personal-name.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+reusing smartphone's addressbook to store personal names of contacts. And moving from that messengers to Matrix confuses
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:23]:
```suggestion
reusing smartphone's addressbook to store personal names of contacts. Moving from these messengers to Matrix is confusing
```

File: proposals/3015-room-personal-name.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+the people, because they can't rename personal chats in his own list like before.
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:24]:
```suggestion
because they can't add personal names to chats like before.
```

File: proposals/3015-room-personal-name.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+item with key `m.room_name_personal` to store custom name of any room for each user individually, with filling personal
+name in content with `room_name_personal` type:
+
+```json
+{
+  "room_name_personal": "Alice Liddell"
+}
+```
+
+By default this item is absent. It is added only when user make the personal renaming of room, and cleared if user
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:39]:
```suggestion
By default this item is absent. It is added only when user sets a personal name for the room, and cleared if user
```

File: proposals/3015-room-personal-name.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+remove the personal name for room (or make it empty).
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:40]:
```suggestion
removes the personal name for the room (or make it empty).
```

Comment (kevincox) [inline proposals/3015-room-personal-name.md:40]: I don't think we should support "making it empty" as a feature. This can be a client UX decision that an empty room name is either not allowed or translates to clearing the personal name on the matrix side.

File: proposals/3015-room-personal-name.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+Regarding to spec, the account data's key can't be deleted, so if user wants to clean the personal name or "Reset to
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:42]:
```suggestion
According to the spec, account data keys can't be deleted, so if user wants to clean the personal name or "Reset to
```

File: proposals/3015-room-personal-name.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+default", the value of the `m.room_name_personal` key should become empty (`{}`).
+
+# Client support
+
+## Displaying:
+
+When client displays the room in list, it should lookup the `m.room_name_personal` key, if it exists and have not empty
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:49]:
```suggestion
When client displays the room in list, it should lookup the `m.room_name_personal` key, if it exists and has a `room_name_personal`
```

File: proposals/3015-room-personal-name.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+value - use it's value for display name in room list instead of global room name. In room page header (over timeline)
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:50]:
```suggestion
attribute - use it's value for display name in room list instead of global room name.
```

File: proposals/3015-room-personal-name.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+clients may show both names (global and personal) with explaining that this room have alternative personal name, that is
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:51]:
```suggestion
Clients may show both names (global and personal) and explainin that this room has an alternative personal name, that is
```

File: proposals/3015-room-personal-name.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+seen only for current user.
+
+## Adding, removing, renaming:
+
+Clients should provide the way to privately rename room and clearly explain the difference between global room name and
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:56]:
```suggestion
Clients should provide a way to privately rename a room and clearly explain the difference between global room name and
```

File: proposals/3015-room-personal-name.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+# Server support
+
+This MSC does not need any changes on server side.
+
+# Potential issues
+
+1. Users can set personal name for room identical as global name of room - this will give little negative effect: all
+   further global room name changes will not change room name in client list. This can be solved via clearly describe
+   the difference between global room name and personal name, and some warning that those names are identical with
+   ability to remove personal name.
```
Comment (kevincox) [inline proposals/3015-room-personal-name.md:75]: I would reword this slightly. Something like.

1. Users can set a personal name that is similar or identical to the global room name. This may cause confusion as the client will not see future global name changes. Clients should consider providing the user a warning and explaining how personal and global room names work.

**Review by kevincox — APPROVED** [review]
Looks good. Just some slight clarifications (mostly removing unnecessary prescriptions for client UX) and a lot of grammar fixes.

File: proposals/3015-room-personal-name.md (hunk 18 above)
Comment (MurzNN) [reply proposals/3015-room-personal-name.md]: Sorry for my English and thanks for all your suggestions! Hope I'm fixed all of them now ;-)

Comment (kevincox) [reply proposals/3015-room-personal-name.md]: No worries. Thanks for the proposal 🙂

**Comment by MurzNN** [comment]
After much thought I decide to rework this MSC to allow override not only room name, but avatar, topic and maybe other room state values too, also reworked the storage format to match the state event content.

File: proposals/3015-room-personal-overrides.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+Most of other modern messengers (Telegram, Skype, Viber, WhatsApp) already have this feature, but only for DM rooms via
+reusing smartphone's addressbook to store personal names of contacts. And moving from that messengers to Matrix confuses
+the people, because they can't rename personal chats in his own list like before.
+
+# Proposal
+
+For solve this problem I propose to use [room's account
+data](https://matrix.org/docs/spec/client_server/r0.6.0#put-matrix-client-r0-user-userid-rooms-roomid-account-data-type)
+item with same key as overriden state event type, but with `override.` prefix, to store personal overrides of needed
```
Comment (emorrp1) [inline proposals/3015-room-personal-overrides.md:33]: since this is account data, is there a need for the `override.` prefix?

Comment (MurzNN) [reply proposals/3015-room-personal-overrides.md:33]: I think no, because we are overriding public data, so we need to force understanding it.

**Review by KitsuneRal — CHANGES_REQUESTED** [review]
I think this is heading in the right direction - but can we please not introduce a new top-level namespace just for the sake of local overriding?

File: proposals/3015-room-personal-overrides.md [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+     "w": 394
+  },
+  "url": "mxc://example.org/JWEIFJgwEIhweiWJE"
+}
+```
+
+By default those items are absent. They are added only when user make the personal override, and cleared if user
+remove the personal name for room (or make it empty). Regarding to spec, the account data's key can't be deleted,
+so if user wants to remove the personal override (aka "Reset to default"), the value of the key should become 
+empty (`{}`).
```
[the remark this answers is not in the document]
Comment (MurzNN) [reply proposals/3015-room-personal-overrides.md:58]: For this value for `override.m.room.name` must be: 
```json
{
  "name": ""
}
```
like regular room state item, when room admin removes the name of room.

File: proposals/3015-room-personal-overrides.md [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+the people, because they can't rename personal chats in his own list like before.
+
+# Proposal
+
+For solve this problem I propose to use [room's account
+data](https://matrix.org/docs/spec/client_server/r0.6.0#put-matrix-client-r0-user-userid-rooms-roomid-account-data-type)
+item with same key as overriden state event type, but with `override.` prefix, to store personal overrides of needed 
+state any room for each user individually, and overriden content, here is example for room name and avatar overrides:
+
+**`override.m.room.name`**
```
[the remark this answers is not in the document]
Comment (MurzNN) [reply proposals/3015-room-personal-overrides.md:36]: By making an new top-level namespace, I tried to separate overrides from other groups of namespaces, to not mix different things in one namespace. But if top level namespace is bad idea, I can rework this to use suffix instead of prefix...

File: proposals/3015-room-personal-overrides.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (KitsuneRal) [reply proposals/3015-room-personal-overrides.md:58]: Yeah, I guess it's worth putting that in the MSC text, for clarity.

File: proposals/3015-room-personal-overrides.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (KitsuneRal) [reply proposals/3015-room-personal-overrides.md:36]: The primary problem with adding a top-level namespace here is that we use reverse-DNS notation, and can't guarantee that `override` won't end up being a legit DNS TLD. More formally, it would probably be the second occasion when we use a non-`m.`, non-revDNS top-level namespace, and I'd rather follow the existing overarching convention. If you're not quite happy with `override` as a postfix, `m.override.room.*` (and `org.matrix.msc3015.override.room.*` for the unstable form - mind that `m.` can be replaced with the unstable prefix in this case) should probably be fine too.

(For the record, the first non-`m.` top-level namespace was `u.`, for room tags; I don't advise employing it here since `u.` tags are defined at the discretion of end users, not the spec.)

File: proposals/3015-room-personal-overrides.md (hunk 19 above)
Comment (uhoreg) [reply proposals/3015-room-personal-overrides.md:33]: I'm resolving this thread because it's essentially the same as the thread below.

File: proposals/3015-room-personal-overrides.md [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+so if user wants to remove the personal override (aka "Reset to default"), the value of the key should become 
+empty (`{}`).
+
+Those overrides can be used for all types of rooms: DMs (to rename personal contacts or set custom photo to avatar, 
+add notes to topic), public rooms (to set desired personal name, eg in user's native language), public Spaces, etc.
+
+For not allow to override bad things, I think we must define an allowlist of state types, that can be overriden:
+- `m.room.name`
+- `m.room.avatar`
+- `m.room.topic`
```
Comment (uhoreg) [inline proposals/3015-room-personal-overrides.md:66]: Room topic is generally a message from the room admins, so I don't think it's really something that people should be overriding.  (Similar for message pinning.)  I can see the value in being able to add your own notes on a room and bookmark certain messages in a room, but I think those should be separate information, in addition to the topic and pinned messages, rather than overriding those.

**Comment by c7hm4r** [comment]
I think it would be better (IMHO essential) to have a custom `m.room.name` only **encrypted** on the server. [It might contain highly sensitive data.](https://github.com/matrix-org/matrix-doc/issues/690#issuecomment-1024953399)

**Comment by Zocker1999NET** [comment]
Hey @MurzNN, because I did not know this MSC exists, I began writing [MSC3864](https://github.com/matrix-org/matrix-spec-proposals/pull/3864) and [MSC3865](https://github.com/matrix-org/matrix-spec-proposals/pull/3865), which kind of overlap with this MSC.

I separated overriding room attributes and user attributes, so overriding user attributes can happen on a global scope and a room scope ([got the idea from a discussion](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$XbIjFaQP-U9uJU2uujULDYWbWUagcKAg64BS3eZC8EI?via=matrix.org&via=libera.chat&via=element.io)). However, as I see it, yours seems more reviewed and so more ready to implement. So I'm unsure if both approaches shall be decided upon or if we could merge both. If you (or others as well) have time to discuss this, I created a public room for that: [#mscs-3015-3864-3865:matrix.org](https://matrix.to/#/#mscs-3015-3864-3865:matrix.org)

**Comment by Fmstrat** [comment]
This seems to have stalled, is there a reason? I'm really surprised with the popularity of the Discord bridge that people don't have more use for this since you end up with 10 rooms all named `#general`.

**Comment by alphapapa** [comment]
> This seems to have stalled, is there a reason? I'm really surprised with the popularity of the Discord bridge that people don't have more use for this since you end up with 10 rooms all named `#general`.

From what I can tell, the next step is to edit the proposal to not use a top-level `override.` namespace for the data, as mentioned by @KitsuneRal at https://github.com/matrix-org/matrix-spec-proposals/pull/3015#pullrequestreview-699603731.

**Comment by alphapapa** [comment]
FWIW, I've implemented this in Ement.el: https://github.com/alphapapa/ement.el/commit/7c12e04d769bc8696bec89730729ac12083aeee3  It seems to work nicely.  Finally I can give rooms names that are meaningful to me.  I hope this proposal will be merged so all clients can use it.

**Comment by lord-ne** [comment]
Is @MurzNN interested in continuing to push the PR forwards? Alternatively, is anyone else interested in picking it up? (Or forking it? I don't know the proper procedure here). This PR has been open for a while (seems like it's even targeting a stale branch), so would it need a lot more changes to be viable?

File: proposals/3015-room-personal-overrides.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (Mart-Bogdan) [reply proposals/3015-room-personal-overrides.md:58]: What would happen if user override it as empty string?

**Comment by Mart-Bogdan** [comment]
There were discussion chat created by @Zocker1999NET [#mscs-3015-3864-3865:matrix.org](https://matrix.to/#/#mscs-3015-3864-3865:matrix.org)

we can jump in.

I would really like it to be moved through.

File: proposals/3015-room-personal-overrides.md [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+available even if user have no permission to change the room name, and same for avatar. If personal value is filled,
+it should be shown near the global room value, with ability to remove it.
+
+# Server support
+
+This MSC does not need any changes on server side.
+
+# Potential issues
+
+1. User can set a personal value that is identical to the current global room value. This may cause confusion as the
```
Comment (Mart-Bogdan) [inline proposals/3015-room-personal-overrides.md:102]: We can store inside our metadata object copy of original value.
And later client app can compare it with current value for room (or user if that 1on1 room) and if it changed -- render exclamation mark.

**Comment by Johennes** [comment]
Given that this proposal appears to be abandoned, gematik have sponsored and are planning to use https://github.com/matrix-org/matrix-spec-proposals/pull/4431 as a (partial) alternative.

**Comment by nvnvnvnvnvnvnv** [comment]
Client implementation ([Voyage](https://github.com/VoyageClient/Voyage)):

https://github.com/VoyageClient/Voyage/commit/53a1fa64263f144087c27397f861ae43c162be5a
